### PR TITLE
STRF-6707 Don't use arguments in helpers

### PR DESCRIPTION
**Pull request by bc-williamkwon**
## What? Why?
STRF 6707 Don't use arguments in helpers.  Move away from arguments pattern so that performance may be improved.

## How was it tested?

----
Added unit test for for helper fix and current unit tests don't break.
cc @bigcommerce/storefront-team

### CODE REVIEW

File: helpers/any.js [hunk 1]
```diff
@@ -9,20 +9,11 @@ const _ = require('lodash');
...
-            any;
-
-        // Translate arguments to array safely
-        for (var i = 0; i < arguments.length; i++) {
-            args.push(arguments[i]);
-        }
-
+    return function(...args) {
+        let any;
         // Take the last argument (content) out of testing array
```
Comment (mattolson) [inline helpers/any.js]: This comment doesn't seem to make sense anymore...

File: helpers/for.js [hunk 2]
```diff
@@ -3,29 +3,30 @@
...
+            options = context;
             if (_.isObject(to)) {
                 context = to;
                 to = from;
                 from = 1;
             }
         }
 
-        if (to <= from) {
+        if (to < from) {
```
Comment (mattolson) [inline helpers/for.js]: Why was this changed? Seems like this will alter current looping logic.

File: helpers/or.js [hunk 3]
```diff
@@ -9,21 +9,12 @@ const _ = require('lodash');
...
-            opts,
-            any;
-
-        // Translate arguments to array safely
-        for (var i = 0; i < arguments.length; i++) {
-            args.push(arguments[i]);
-        }
-
+    return function(...args) {
         // Take the last argument (content) out of testing array
```
Comment (mattolson) [inline helpers/or.js]: This comment seems incorrect now

File: helpers/for.js (hunk 2 above)
Comment (bc-williamkwon) [reply helpers/for.js]: It allows for one iteration.  Previously we could only do two iterations at the minimum.  There was an issue written about it too.

File: helpers/or.js (hunk 3 above)
Comment (bc-williamkwon) [reply helpers/or.js]: I thought it made sense as arguments (non-keyword) are passed to functions.

File: helpers/for.js (hunk 2 above)
Comment (mattolson) [reply helpers/for.js]: Let's pull this out as a separate change so we can evaluate it separately.

File: helpers/or.js (hunk 3 above)
Comment (mattolson) [reply helpers/or.js]: Maybe if it said "options" instead of "content", it would make more sense.

Comment (bc-williamkwon) [reply helpers/or.js]: fixed

Comment (bc-williamkwon) [reply helpers/or.js]: Nvm rebasing issues . . .

Comment (bc-williamkwon) [reply helpers/or.js]: Git is behaving weird but I believe things are fixed now.

Comment (bc-williamkwon) [reply helpers/or.js]: whitespace in test case of for.js is fixed in the other pr.

Comment (bc-williamkwon) [reply helpers/or.js]: fixed here now too actually.

Comment (bc-williamkwon) [reply helpers/or.js]: Actually realized that I missed a file.

Comment (bc-williamkwon) [reply helpers/or.js]: Three files have been changed to be more explicit about what is being pulled out of the args array.

**Comment by bc-williamkwon** [comment]
The rest parameter pattern (...args) is more performant when processing things as an array but otherwise it's better to use named parameters.  I changed contains and stylesheet based on documentation ```contains Helper
The contains helper is custom to Stencil. It checks whether the second parameter is included in the first parameter (typically a collection).```
```{{stylesheet}}
The stylesheet helper is custom to Stencil. It renders a link tag to insert a stylesheet into your theme. (This is required if you want Theme Editor to rewrite the stylesheet file when a merchant customizes their theme.) This helper returns an HTML string.

Parameters
path: String containing the path to the theme’s CSS stylesheet file.
Other parameters are optional, appended in the form: key="value".
Example
{{{stylesheet "assets/css/style.css" class="myStylesheet"}}}```

Also partial.js is called in the form of Handlebars.registerPartial('myPartial', '{{name}}') so I think this should be safe as well.

File: helpers/all.js [hunk 4]
```diff
@@ -9,19 +9,12 @@ const _ = require('lodash');
  * {{#all items theme_settings.optionA theme_settings.optionB}} ... {{/all}}
  */
 const factory = () => {
-    return function() {
```
Comment (mattolson) [inline helpers/all.js]: I'm going to start leaving comments for each helper individually to help me keep track of what portions are approved or still under review.

Changes to this helper: approved.

File: helpers/any.js [hunk 5]
```diff
@@ -9,20 +9,11 @@ const _ = require('lodash');
  * {{#any items selected=true}} ... {{/any}}
  */
 const factory = () => {
-    return function() {
```
Comment (mattolson) [inline helpers/any.js]: Changes to this helper: approved

File: helpers/compare.js [hunk 6]
```diff
@@ -1,19 +1,14 @@
 'use strict';
 
 const factory = () => {
-    return function(lvalue, rvalue) {
```
Comment (mattolson) [inline helpers/compare.js]: Changes to this helper: approved

File: helpers/block.js [hunk 7]
```diff
@@ -1,10 +1,9 @@
 'use strict';
 
 const factory = globals => {
-    return function(name) {
```
Comment (mattolson) [inline helpers/block.js]: Changes to this helper: approved

File: helpers/contains.js [hunk 8]
```diff
@@ -10,11 +10,8 @@ const _ = require('lodash');
  * {{#contains font_path "Roboto"}} ... {{/contains}}
  */
 const factory = () => {
-    return function() {
```
Comment (mattolson) [inline helpers/contains.js]: Changes to this helper: approved

File: helpers/for.js [hunk 9]
```diff
@@ -3,21 +3,22 @@
 const _ = require('lodash');
 
 const factory = () => {
-    return function(from, to, context) {
```
Comment (mattolson) [inline helpers/for.js]: Changes to this helper: approved

File: helpers/join.js [hunk 10]
```diff
@@ -1,9 +1,8 @@
 'use strict';
 
 const factory = () => {
-    return function(array, separator) {
```
Comment (mattolson) [inline helpers/join.js]: Changes to this helper: approved

File: helpers/lang.js [hunk 11]
```diff
@@ -1,8 +1,7 @@
 'use strict';
 
 const factory = globals => {
-    return function(translationKey) {
```
Comment (mattolson) [inline helpers/lang.js]: Changes to this helper: approved

File: helpers/unless.js [hunk 12]
```diff
@@ -1,15 +1,15 @@
 'use strict';
 
 const factory = globals => {
-    return function() {
```
Comment (mattolson) [inline helpers/unless.js]: Changes to this helper: approved

File: helpers/stylesheet.js [hunk 13]
```diff
@@ -4,13 +4,11 @@ const _ = require('lodash');
 const buildCDNHelper = require('./lib/cdnify');
 
 const factory = globals => {
-    return function(assetPath) {
+    return function(assetPath, options) {
```
Comment (mattolson) [inline helpers/stylesheet.js]: This doesn't allow for html attributes that don't have values (i.e. not passed in via options.hash)

File: helpers/replace.js [hunk 14]
```diff
@@ -1,9 +1,7 @@
 'use strict';
 
 const factory = () => {
-    return function(needle, haystack) {
```
Comment (mattolson) [inline helpers/replace.js]: Changes to this helper: approved

File: helpers/partial.js [hunk 15]
```diff
@@ -1,8 +1,8 @@
 'use strict';
 
 const factory = globals => {
-    return function(name) {
```
Comment (mattolson) [inline helpers/partial.js]: Changes to this helper: approved

File: helpers/or.js [hunk 16]
```diff
@@ -9,21 +9,12 @@ const _ = require('lodash');
  * {{#or 1 0 0 0 0 0}} ... {{/or}}
  */
 const factory = () => {
-    return function() {
```
Comment (mattolson) [inline helpers/or.js]: Changes to this helper: approved

File: helpers/if.js [hunk 17]
```diff
@@ -3,8 +3,7 @@
 const _ = require('lodash');
 
 const factory = globals => {
-    return function(lvalue, operator, rvalue) {
```
Comment (mattolson) [inline helpers/if.js]: Changes to this helper: approved

File: helpers/getFontsCollection.js [hunk 18]
```diff
@@ -3,9 +3,8 @@
 const getFonts = require('./lib/fonts');
 
 const factory = globals => {
-    return function() {
```
Comment (mattolson) [inline helpers/getFontsCollection.js]: Changes to this helper: approved

File: spec/helpers/stylesheet.js [hunk 19]
```diff
@@ -25,7 +25,7 @@ describe('stylesheet helper', () => {
     it('should render a link tag and all extra attributes with no cdn url', done => {
         runTestCases([
             {
-                input: '{{{stylesheet "assets/css/style.css" blah rel="something" class="myStyle"}}}',
```
[the remark this answers is not in the document]
Comment (bc-williamkwon) [reply spec/helpers/stylesheet.js]: ok, thanks for your insight.  I will fix it.

**Comment by bc-williamkwon** [comment]
rebased and fixed merge conflicts.

File: helpers/stylesheet.js [hunk 20]
```diff
@@ -4,13 +4,11 @@ const _ = require('lodash');
 const buildCDNHelper = require('./lib/cdnify');
 
 const factory = globals => {
-    return function(assetPath) {
```
Comment (mattolson) [inline helpers/stylesheet.js]: Changes to this helper: approved

**Comment by bc-williamkwon** [comment]
Thanks for doing the code review.
